File: src/main/java/br/com/alura/forum/security/controller/UserAuthenticationController.java
package br.com.alura.forum.security.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import br.com.alura.forum.controller.dto.input.LoginInputDto;
import br.com.alura.forum.controller.dto.output.AuthenticationTokenOutputDto;
import br.com.alura.forum.security.jwt.TokenManager;

@RestController
@RequestMapping("/api/auth")
public class UserAuthenticationController {

	@Autowired
	private AuthenticationManager authManager;
	
	@Autowired
	private TokenManager tokenManager;

	@PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, 
			produces = MediaType.APPLICATION_JSON_VALUE)
	public ResponseEntity<AuthenticationTokenOutputDto> authenticate(@RequestBody LoginInputDto loginInfo) {
	
		UsernamePasswordAuthenticationToken authenticationToken = loginInfo.build();

		try {
			Authentication authentication = authManager.authenticate(authenticationToken); 			
			String jwt = tokenManager.generateToken(authentication);
			
			AuthenticationTokenOutputDto tokenResponse = new AuthenticationTokenOutputDto("Bearer", jwt);
			return ResponseEntity.ok(tokenResponse);
		
		} catch (AuthenticationException e) {
			return ResponseEntity.badRequest().build();
		}
		
	}
}

File: src/main/kotlin/br/com/alura/forum/security/controller/UserAuthenticationController.kt
package br.com.alura.forum.security.controller

import br.com.alura.forum.controller.dto.input.LoginInputDto
import br.com.alura.forum.controller.dto.output.AuthenticationTokenOutputDto
import br.com.alura.forum.security.jwt.TokenManager
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.core.AuthenticationException
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("api/auth")
class UserAuthenticationController(private val authManager: AuthenticationManager,
                                   private val tokenManager: TokenManager) {

    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE],
            produces = [MediaType.APPLICATION_JSON_VALUE])
    fun authenticate(@RequestBody loginInfo: LoginInputDto): ResponseEntity<AuthenticationTokenOutputDto> {
        val authenticationToken = loginInfo.build()

        return try {
            val authentication = authManager.authenticate(authenticationToken)
            val jwtToken = tokenManager.generateToken(authentication)

            val tokenOutputDto = AuthenticationTokenOutputDto("Bearer", jwtToken)

            ResponseEntity.ok(tokenOutputDto)
        } catch (e: AuthenticationException) {

            ResponseEntity.badRequest().build()
        }
    }
}